skopt/acquisition.py

```diff
@@ -264,3 +266,109 @@ def gaussian_ei(X, model, y_opt=0.0, xi=0.01, return_grad=False):
         return values, grad
 
     return values
+
+
+def gaussian_entropy_search(X, model, n_rep_points=50,
+                            n_trial_points=200, measure="EI", n_samples=500,
+                            random_state=None, rep_cands=None, **kwargs):
+    """
+    Use entropy search to optimize the acquisition function.
+
+    1. It approximates the search space by a set of representer points sampled from
+       a suitable measure such as the "probability of improvement" or
+       "expected improvement".
+
+    2. p_min at a representer point X_i is calculated by the Integral
+       p(f(x)) \prod_{j=1}^N_rep \Theta(f(X_j) - f(X_i)) \\
+       Where P(f(x)) is the GP posterior and \Theta is the heaviside step function.
+       This can be used to calculate the base entropy.
+
+    3. Then for each candidate point, the change in the mean and the covariance
+       of each of these representer points can be computed. This in turn
+       can be used to compute the change in p_min for each of these representer
+       points.
+
+    Parameters
+    ----------
+    * `X` [array-like, shape=(n_samples, n_features)]:
+        Values where the acquisition function should be computed.
+
+    * `model` [sklearn estimator that implements predict with ``return_std``]:
+        The fit estimator that approximates the function through the
+        method ``predict``.
+        It should have a ``return_std`` parameter that returns the standard
+        deviation.
+
+    * `space` [Space instance]:
+        This is required to sample representer points.
+
+    * `n_rep_points` [int, default 50]:
+        Number of representer points.
+
+    * `n_trial_points` [int, default 200]:
+        Number of trial points sampled uniformly for each representer point
+        candidate.
+
+    * `n_samples` [int, default 500]:
+        Number of functions to sample from the GP posterior at the
+        representer points to approximate p_min at each of the representer
+        points.
+
+    * `random_state` [int, RandomState instance, or None (default)]:
+        For reproducible results.
+    """
+    rng = np.random.RandomState(random_state)
+    n_features = X.shape[1]
+
+    # Step 1: Sample `n_representer_points` from the given measure.
+    # XXX: Use MCMC sampling to sample from the quadrature measure.
+    if measure == "LCB":
+        quad_measure = gaussian_lcb
+
+    elif measure in ["EI", "PI"]:
+        if measure == "EI":
+            quad_measure = lambda X : -gaussian_ei(X, model)
+        else:
+            quad_measure = lambda X : -gaussian_pi(X, model)
+
+    n_total = n_rep_points * n_trial_points
+    all_values = np.reshape(quad_measure(rep_cands), (n_rep_points, n_trial_points))
+    rep_inds = np.arange(0, n_total, n_trial_points) + np.argmin(all_values, axis=1)
+    rep_points = rep_cands[rep_inds]
```

Comment:
It should be set outside the loop, since this should be constant for every acquisition function call.